File: php/phpsearch/fileutil.php
<?php

class FileUtil {

    public static $DOT_PATHS = array('.', '..');

    public static function expand_user_home_path(string $path): string {
        if (strpos($path, '~') === 0) {
            return str_replace('~', getenv('HOME'), $path);
        }
        return $path;
    }

    public static function get_extension(string $file): string {
        $f = basename($file);
    	$ext = '';
    	$dot_idx = strrpos($f, '.');
    	if ($dot_idx !== false && $dot_idx > 0 && $dot_idx < strlen($f))
    		$ext = substr($f, $dot_idx+1);
        return $ext;
    }

    public static function is_dot_dir(string $d): bool {
        return in_array($d, self::$DOT_PATHS);
    }

    public static function is_hidden(string $file): bool {
        $f = basename($file);
        return strlen($f) > 1 && $f{0}==='.' && !self::is_dot_dir($f);
    }

    public static function join_path(string $path, string $file) {
        return self::normalize_path($path) . self::get_separator($path) . $file;
    }

    public static function get_separator(string $path): string {
        $sep = '/';
        if (strpos($path, $sep) === false) {
            if (strpos($path, '\\') === false) {
                return $sep;
            } else {
                return '\\';
            }
        }
        return $sep;
    }

    public static function normalize_path(string $path): string {
        $sep = self::get_separator($path);
        return rtrim($path, $sep);
    }

    public static function split_path(string $path) {
        $sep = self::get_separator($path);
        if ($sep == '/')
            $sep = '\\/';
        return preg_split("/$sep/", $path);
    }
}

?>

File: cpp/cppsearch/src/FileUtil.cpp
#include <boost/algorithm/string.hpp>
#include <boost/filesystem.hpp>
#include <sys/stat.h>
#include <fstream>
#include "FileUtil.h"

namespace cppsearch {
    // TODO: find lib function for this, current implementation is incomplete
    std::string FileUtil::expand_path(const std::string& file_path) {
        if (file_path.at(0) == '~') {
            std::string expanded = getenv("HOME");
            if (file_path.length() > 1) {
                expanded.append(file_path.substr(1));
            }
            return expanded;
        }
        return file_path;
    }

    bool FileUtil::file_exists(const std::string& name) {
        struct stat st;
        return (stat(name.c_str(), &st) == 0);
    }

    uint64_t FileUtil::file_size(const std::string& file_path) {
        struct stat st;
        if (stat(file_path.c_str(), &st)) /*failure*/
            return -1; // when file does not exist or is not accessible
        return (uint64_t) st.st_size;
    }

    std::string FileUtil::get_contents(const std::string& file_path) {
        std::ifstream fin(file_path);
        std::string contents = get_contents(fin);
        fin.close();
        return contents;
    }

    std::string FileUtil::get_contents(const std::ifstream& fin) {
        std::stringstream sstr;
        sstr << fin.rdbuf();
        return sstr.str();
    }

    std::string FileUtil::get_extension(const std::string& name) {
        boost::filesystem::path path(name);
        std::string ext = path.extension().string();
        if (name == ext) {
            return "";
        }
        if (!ext.empty() && ext[0] == '.') {
            ext = ext.substr(1);
        }
        if (ext != "Z") {
            ext = boost::to_lower_copy(ext);
        }
        return ext;
    }

    std::string FileUtil::get_file_name(const std::string& file_path) {
        boost::filesystem::path path(file_path);
        std::string file_name = path.filename().string();
        return file_name;
    }

    bool FileUtil::is_directory(const std::string& name) {
        boost::filesystem::path path(name);
        return boost::filesystem::is_directory(path);
    }

    bool FileUtil::is_regular_file(const std::string& name) {
        boost::filesystem::path path(name);
        return boost::filesystem::is_regular_file(path);
    }

    bool FileUtil::is_dot_dir(const std::string& name) {
        return name == "." || name == ".." ||
            name == "./" || name == "../" ||
            name == ".\\" || name == "..\\";
    }

    bool FileUtil::is_hidden(const std::string& name) {
        boost::filesystem::path path(name);
        return !name.empty() && name.at(0) == '.' && !FileUtil::is_dot_dir(name);
    }

    std::string FileUtil::join_path(const std::string& path1, const std::string& path2) {
        boost::filesystem::path p1(path1);
        boost::filesystem::path p2(path2);
        boost::filesystem::path fullpath = p1 / p2;
        return fullpath.string();
    }

    std::vector<std::string> FileUtil::split_path(const std::string& file_path) {
        std::vector<std::string> parts;
        boost::filesystem::path path(file_path);
        if (FileUtil::is_dot_dir(file_path)) {
            if (path.filename_is_dot()) {
                parts.emplace_back(".");
            } else if (path.filename_is_dot_dot()) {
                parts.emplace_back("..");
            }
            return parts;
        }
        int i = 0;
        for(auto& part : path) {
            if (part.string() == ".") {
                if (i == 0) {
                    parts.push_back(part.string());
                }
            } else if (part.string() != "/") {
                parts.push_back(part.string());
            }
            i++;
        }
        return parts;
    }
}
